transport/grpc/header.go

```diff
@@ -0,0 +1,86 @@
+package grpc
+
+import (
+	"strings"
+
+	"go.uber.org/yarpc/transport"
+
+	"google.golang.org/grpc/metadata"
+)
+
+// headerMapper converts gRPC Metadata to and from transport headers.
+type headerMapper struct{ Prefix string }
+
+var (
+	applicationHeaders = headerMapper{ApplicationHeaderPrefix}
+)
+
+// toMetadata converts transport headers into gRPC Metadata.
+//
+// Headers are read from 'from' and written to 'to'. The final header collection
+// is returned.
+//
+// If 'to' is nil, a new map will be assigned.
+func (hm headerMapper) toMetadata(from transport.Headers, to metadata.MD) metadata.MD {
+	if to == nil {
+		to = make(metadata.MD, from.Len())
+	}
+	for k, v := range from.Items() {
+		headers(to).add(hm.Prefix+k, v)
+	}
+	return to
+}
+
+// fromMetadata converts GRPC Metadata to transport headers.
+//
+// Headers are read from 'from' and written to 'to'. The final header collection
+// is returned.
+//
+// If 'to' is nil, a new map will be assigned.
+func (hm headerMapper) fromMetadata(from metadata.MD, to transport.Headers) transport.Headers {
+	prefixLen := len(hm.Prefix)
+	for k := range from {
+		if strings.HasPrefix(k, hm.Prefix) {
+			key := k[prefixLen:]
+			to = to.With(key, headers(from).get(k))
+		}
+		// Note: undefined behavior for multiple occurrences of the same header
+	}
+	return to
+}
+
+// headers is a convenience type for converting Header information safely
```

Comment:
You say 'safely' but you're not normalizing the keys. Don't you need to do that?